File: Applied Course/4.Problem Solving/12.Problems on Backtracking/5.Permutations.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
public:
    vector<vector<int>> result;
    void permutations(vector<int>& nums,int l,int r)
    {
        if(l==r-1)
            result.push_back(nums);
        else
        {
            for(int i=l;i<r;i++)
            {
                swap(nums[l],nums[i]);
                permutations(nums,l+1,r);
                swap(nums[l],nums[i]);
            }
            
        }
    }  
    vector<vector<int>> permute(vector<int>& nums) {
        int n=nums.size();
        permutations(nums,0,n);
        return result;
    }
};
int main()
{
    // https://leetcode.com/problems/permutations/
    return 0;
}

File: Applied Course/4.Problem Solving/12.Problems on Backtracking/5.Permutations.py
# https://leetcode.com/problems/permutations
class Solution:
    def __init__(self):
        self.result=[]
    def permutations(self,nums,l,r):
        if l==r-1:
            self.result.append(nums[:])
        else:
            for i in range(l,r):
                nums[l],nums[i]=nums[i],nums[l]
                self.permutations(nums,l+1,r)
                nums[l],nums[i]=nums[i],nums[l]
    
    def permute(self, nums: List[int]) -> List[List[int]]:
        n=len(nums)
        self.permutations(nums,0,n)
        return self.result